ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -654,34 +677,37 @@ private void updateVersionStateAndDeleteDatasetVersionOutOfRetentionCount() thro
-          //In the last callback, set the rest method back.
-          LOGGER.debug("Change rest method back to PUT");
-          restRequest.setRestMethod(RestMethod.PUT);
+          //make sure to process response delete is done.
+          processResponseCallback.onCompletion(null, e);
         };
       }
       Callback<Void> nextCallBack =
-          recursiveCallback(datasetVersionRecordList, idx + 1, accountName, containerName, datasetName);
+          recursiveCallback(datasetVersionRecordList, idx + 1, accountName, containerName, datasetName,
+              processResponseCallback, null);
       DatasetVersionRecord record = datasetVersionRecordList.get(idx);
       return buildCallback(frontendMetrics.deleteBlobSecurityProcessResponseMetrics, securityCheckResult -> {
-        String version = record.getVersion();
-        RequestPath requestPath = getRequestPath(restRequest);
-        RequestPath newRequestPath =
-            new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
-                NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH + version,
-                requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
-        LOGGER.debug("New request path in recursive call : " + newRequestPath);
-        // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
-        restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
-        restRequest.setArg(InternalKeys.TARGET_ACCOUNT_KEY, null);
-        restRequest.setArg(InternalKeys.TARGET_CONTAINER_KEY, null);
-        deleteBlobHandler.handle(restRequest, restResponseChannel, nextCallBack);
-      }, uri, LOGGER, (r, e) -> LOGGER.error("Delete dataset" + restRequest.getUri() + " failed with exception" + e));
+            String version = record.getVersion();
+            RequestPath requestPath = getRequestPath(wrappedRestRequest);
+            RequestPath newRequestPath =
+                new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                    NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH + version,
+                    requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+            LOGGER.debug("New request path in recursive call : " + newRequestPath);
+            // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+            wrappedRestRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+            wrappedRestRequest.setArg(InternalKeys.TARGET_ACCOUNT_KEY, null);
+            wrappedRestRequest.setArg(InternalKeys.TARGET_CONTAINER_KEY, null);
+            deleteBlobHandler.handle(wrappedRestRequest, restResponseChannel, nextCallBack);
+          }, uri, LOGGER,
+          (r, e) -> LOGGER.error("Delete dataset" + wrappedRestRequest.getUri() + " failed with exception" + e));
```

Comment:
this should not just be a logger, this callback will be invoked when the delete failed, we should still call processResponseCallback.onComplete to return 201 back to client.

Reply:
can you please add a test when the delete handler fails, client would still get a success response.